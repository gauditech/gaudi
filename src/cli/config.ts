import path from "path";

import { EngineConfig } from "@src/config";

// -------------------- Various configs, constants and defaults

/**
 * Time to wait before reporting resource watcher changes (in millis).
 *
 * This allows basic debouncing (eg. when creating/updating multiple files at once).
 */
export const RESOURCE_WATCH_DELAY = 500;

/** Returns path to prisma schema file */
export function getDbSchemaPath(config: EngineConfig): string {
  return `${config.gaudiFolder}/db/schema.prisma`;
}

// defaul Node options
const DEFAULT_NODE_OPTIONS = [
  // development node options - maybe we should allow disabling them in production?
  "--enable-source-maps",
  "--stack-trace-limit=30",
  "--stack-size=2048",
];

/**
 * Default node options
 *
 * Additional node options can be passed via NODE_OPTIONS env var which is included when running commands
 */
export function getDefaultNodeOptions(): string[] {
  return [...DEFAULT_NODE_OPTIONS];
}

/**
 * Paths to Gaudi scripts
 *
 * Target Gaudi scripts directly instead of via NPX because we cannot pass some node options through NPX (via --node-options")
 * See a list of allowed options here: https://nodejs.org/docs/latest-v16.x/api/cli.html#node_optionsoptions
 */
export const GAUDI_SCRIPTS = {
  ENGINE: path.join(__dirname, "../engine.js"),
  RUNTIME: path.join(__dirname, "../runtime/runtime.js"),
  POPULATOR: path.join(__dirname, "../populator/populator.js"),
};

/**
 * Add path to CLI's binary.
 *
 * We don't want to force users to be aware of our binaries which means that
 * they probably won't be available as global binaries thus we have to manually
 * target them in our own `node_modules`.
 */
export function appendBinPath(binName: string): string {
  return path.join("./node_modules/@gaudi/engine/node_modules/.bin/", binName);
}
