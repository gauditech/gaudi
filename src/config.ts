import dotenv from "dotenv";
import { ConnectionOptions } from "pg-connection-string";

import { parseConnectionString } from "./common/utils";

import { GAUDI_FOLDER_NAME } from "@src/const";

export type EngineConfig = {
  /** Path to Gaudi blueprint file */
  inputPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** Gaudi folder */
  gaudiFolder: string;
  dbConn: ConnectionOptions;
  embeddedPg: boolean;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(configPath?: string): EngineConfig {
  dotenv.config({ path: configPath });

  const inputPath = process.env.GAUDI_ENGINE_INPUT_PATH ?? "";
  const outputFolder = process.env.GAUDI_ENGINE_OUTPUT_PATH ?? ".";
  // gaudi folder's path should probably be determined by the position of (future) gaudi config file
  const gaudiFolder = `./${GAUDI_FOLDER_NAME}`;

  const dbConn = parseConnectionString(process.env.GAUDI_DATABASE_URL);
  const embeddedPg = process.env.GAUDI_EMBEDDED_POSTGRESQL_ENABLED?.toLowerCase() === "true";

  return { inputPath, outputFolder, gaudiFolder, dbConn, embeddedPg };
}
