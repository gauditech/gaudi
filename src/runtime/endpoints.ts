import express, { Request } from "express";
import knex from "knex";

const app = express();
const port = 3002;

import {
  PathParam,
  buildEndpointPath,
  queryableFromEndpointTargets,
  selectToSelectable,
} from "@src/builder/query";
import { Definition, QueryDef, QueryDefPath, TargetDef } from "@src/types/definition";

function validatePathParam(param: PathParam["params"][number], val: string): PathVariable {
  switch (param.type) {
    case "integer":
      return { name: param.name, type: param.type, value: parseInt(val, 10) };
    case "text":
      return { name: param.name, type: param.type, value: val };
  }
}

type PathVariable =
  | { name: string; type: "text"; value: string }
  | { name: string; type: "integer"; value: number };

export function setupEndpoints(def: Definition) {
  for (const entrypoint of def.entrypoints) {
    for (const endpoint of entrypoint.endpoints) {
      switch (endpoint.kind) {
        case "get": {
          const { path, params } = buildEndpointPath(endpoint);
          app.get(path, (req: Request, res) => {
            // collect path params
            const vars = params.map((p) => validatePathParam(p, req.params[p.name]));

            // get targets
            const q = queryableFromEndpointTargets(
              def,
              endpoint.targets,
              selectToSelectable(endpoint.response),
              "single"
            );
          });
        }
      }
    }
  }
}

/**
 * Here we go again... Returning a query/able, whatever it may be.


from Org.public_repos.issues
filter
      Org.slug is ...
  and public_repos.slug is ...
  and     
*/
