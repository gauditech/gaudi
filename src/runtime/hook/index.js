/** Returns given value. Can be used for testing. */
module.exports.echo = function ({ value }) {
  return value;
};
