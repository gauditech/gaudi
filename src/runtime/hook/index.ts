/** Concatente all given string parts */
export function concat(...parts: string[]): string {
  return parts.join("");
}
