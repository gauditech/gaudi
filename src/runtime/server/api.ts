import path from "path";

import { Express, NextFunction, Request, Response, static as staticHandler } from "express";
import { OpenAPIV3 } from "openapi-types";
import { serve, setup } from "swagger-ui-express";

import { buildOpenAPI } from "@src/builder/openAPI";
import { saveOutputFile } from "@src/common/utils";
import { buildEntrypoints } from "@src/runtime/server/admin";
import { buildEndpoints as buildAuthEndpoints } from "@src/runtime/server/authentication";
import { getAppContext } from "@src/runtime/server/context";
import { buildEndpointConfig, registerServerEndpoint } from "@src/runtime/server/endpoints";
import { EndpointConfig } from "@src/runtime/server/types";
import { Definition, EntrypointDef } from "@src/types/definition";

/** Create endpoint handlers, OpenAPI specs and attach them to server instance */
export function setupServerApis(definition: Definition, app: Express) {
  const config = getAppContext(app).config;

  // --- static folder (eg. for API specs)
  const specOutputFolder = path.join(config.outputFolder, "api-spec");
  app.use("/api-spec", staticHandler(specOutputFolder));

  // --- definition API (aka the "blueprint")
  const definitionEntrypoints = definition.entrypoints;
  const definitionEndpointConfigs = buildEndpointConfig(definition, definitionEntrypoints);
  setupEntrypointApi(
    "api",
    definition,
    definitionEntrypoints,
    definitionEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- admin API
  const adminEntrypoints = buildEntrypoints(definition);
  const adminEndpointConfigs = buildEndpointConfig(definition, adminEntrypoints);
  setupEntrypointApi(
    "api-admin",
    definition,
    adminEntrypoints,
    adminEndpointConfigs,
    app,
    specOutputFolder
  );

  // --- authentication API
  const authConfigs = buildAuthEndpoints();
  setupConfigEndpoints("api", authConfigs, app);
}

/** Create API endpoints from entrypoints.*/
function setupEntrypointApi(
  name: string,
  definition: Definition,
  entrypoints: EntrypointDef[],
  endpointConfigs: EndpointConfig[],
  app: Express,
  outputFolder: string
) {
  const basePath = `/${name}`;
  const specFile = path.join(outputFolder, `${name}.openapi.json`);

  endpointConfigs.forEach((epc) => registerServerEndpoint(app, epc, basePath));

  setupEntrypointApiSpec(definition, entrypoints, app, basePath, specFile);
}

/** Create API OpenAPI spec from entrpyoints */
function setupEntrypointApiSpec(
  definition: Definition,
  entrypoints: EntrypointDef[],
  app: Express,
  basePath: string,
  outputFile: string
) {
  const openApi = buildOpenAPI(
    { models: definition.models, entrypoints, populators: [] },
    basePath
  );

  saveOutputFile(outputFile, JSON.stringify(openApi, undefined, 2));

  setupEntrypointApiSwagger(openApi, app, basePath);
}

function setupEntrypointApiSwagger(
  openApiDocument: OpenAPIV3.Document,
  app: Express,
  apiPath: string
) {
  const swaggerPath = `${apiPath}-docs`;

  app.use(swaggerPath, serve, (_req: Request, _resp: Response, _next: NextFunction) =>
    setup(openApiDocument)(_req, _resp, _next)
  );
}

/** Setup API endpoints from endpoint configs */
function setupConfigEndpoints(name: string, configs: EndpointConfig[], app: Express) {
  const basePath = `/${name}`;

  configs.forEach((epc) => {
    registerServerEndpoint(app, epc, basePath);
  });

  // NOTE: we cannot currently create OpenAPI spec from endpoint configs
}
