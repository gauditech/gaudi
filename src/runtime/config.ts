import fs from "fs";

import { initializeContext } from "@src/runtime/server/context";
import { createDbConn } from "@src/runtime/server/dbConn";
import { Definition } from "@src/types/definition";

export type RuntimeConfig = {
  /** Runtime server host name */
  host: string;
  /** Runtime server port number */
  port: number;
  /** Path to generated definition.json file. */
  definitionPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** DB connection URL */
  dbConnUrl: string;
  /** DB default schema */
  dbSchema?: string;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(): RuntimeConfig {
  const host = process.env.GAUDI_RUNTIME_SERVER_HOST ?? "127.0.0.1";
  const port =
    process.env.GAUDI_RUNTIME_SERVER_PORT != null
      ? parseInt(process.env.GAUDI_RUNTIME_SERVER_PORT, 10)
      : 3001;
  const definitionPath = process.env.GAUDI_RUNTIME_DEFINITION_PATH || "definition.json";
  const outputFolder = process.env.GAUDI_RUNTIME_OUTPUT_PATH || ".";

  const dbConnUrl = process.env.GAUDI_DATABASE_URL || "";
  const dbSchema = process.env.GAUDI_DATABASE_SCHEMA;

  return { host, port, definitionPath, outputFolder, dbConnUrl, dbSchema };
}

/** Load definition file and return it's content */
export function loadDefinition(definitionPath: string): Definition {
  // --- read input file
  if (!fs.existsSync(definitionPath)) {
    throw new Error(`Definition file not found: "${definitionPath}"`);
  }
  const definitionStr = fs.readFileSync(definitionPath).toString("utf-8");
  return JSON.parse(definitionStr);
}

export function createAppContext(config: RuntimeConfig) {
  initializeContext({
    dbConn: createDbConn(config.dbConnUrl, {
      schema: config.dbSchema,
    }),
    config: config,
  });
}
