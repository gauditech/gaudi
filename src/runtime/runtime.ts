// import this file only with relative path because this file actually configures path aliases (eg @src, ...)
import "../common/setupAliases";

import { loadDefinition, readConfig } from "@src/runtime/config";
import { importHooks } from "@src/runtime/hooks";
import { createServer } from "@src/runtime/server/server";

// read environment
const config = readConfig();

const definition = loadDefinition(config.definitionPath);

importHooks(config);

// start server
createServer(definition, config);
