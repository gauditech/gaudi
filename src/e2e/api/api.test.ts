import fs from "fs";
import path from "path";

import _ from "lodash";
import request from "supertest";

import { PopulatorData, createApiTestSetup } from "@src/e2e/api/setup";
import { readConfig } from "@src/runtime/config";

// these tests last longer than default 5s timeout so this seems to help
jest.setTimeout(10000);

describe("API endpoints", () => {
  const config = readConfig(path.join(__dirname, "api.test.env"));

  const { getServer, setup, destroy } = createApiTestSetup(
    config,
    loadBlueprint(path.join(__dirname, "api.model.gaudi")),
    loadPopulatorData(path.join(__dirname, "api.data.json"))
  );

  describe("Org", () => {
    beforeAll(async () => {
      await setup();
    });
    afterAll(async () => {
      await destroy();
    });

    it("get", async () => {
      const response = await request(getServer()).get("/org/org1");

      expect(response.statusCode).toBe(200);
      expect(response.body).toMatchSnapshot();
    });

    it("list", async () => {
      const response = await request(getServer()).get("/org");

      expect(response.statusCode).toBe(200);
      expect(response.body).toMatchSnapshot();
    });

    it("create", async () => {
      const data = {
        name: "Org NEW",
        slug: "orgNEW",
        description: "Org NEW description",
      };
      const postResp = await request(getServer()).post("/org").send(data);

      expect(postResp.statusCode).toBe(200);
      const getResp = await request(getServer()).get("/org/orgNEW");

      expect(getResp.statusCode).toBe(200);
      expect(getResp.body).toMatchSnapshot();

      // ensure `.get` result is the same as returned from `create` method
      expect(getResp.body).toEqual(postResp.body);
    });

    it("update", async () => {
      const data = { slug: "org2", name: "Org 2A", description: "Org 2A description" };

      const patchResp = await request(getServer()).patch("/org/org2").send(data);
      expect(patchResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org2");
      expect(getResp.statusCode).toBe(200);
      expect(getResp.body).toMatchSnapshot();

      // ensure `.get` result is the same as returned from `update` method
      expect(getResp.body).toEqual(patchResp.body);
    });

    it("delete", async () => {
      const patchResp = await request(getServer()).delete("/org/org3");
      expect(patchResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org3");
      expect(getResp.statusCode).toBe(404);
    });
  });

  describe("Repo", () => {
    beforeAll(async () => {
      await setup();
    });
    afterAll(async () => {
      await destroy();
    });

    it("get", async () => {
      const response = await request(getServer()).get("/org/org1/repos/1");

      expect(response.statusCode).toBe(200);
      expect(response.body).toMatchSnapshot();
    });

    it("list", async () => {
      const response = await request(getServer()).get("/org/org1/repos");

      expect(response.statusCode).toBe(200);
      expect(_.sortBy(response.body, "id")).toMatchSnapshot();
    });

    it("create", async () => {
      const data = {
        name: "Repo 6",
        slug: "repo6",
        description: "Repo 6 description",
        is_public: true,
      };
      const postResp = await request(getServer()).post("/org/org1/repos").send(data);
      expect(postResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org1/repos/6");
      expect(getResp.statusCode).toBe(200);
      expect(getResp.body).toMatchSnapshot();
    });

    it("update", async () => {
      const data = { slug: "repo2", name: "Repo 2A", description: "Repo 2A description" };

      const patchResp = await request(getServer()).patch("/org/org1/repos/2").send(data);
      expect(patchResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org1/repos/2");
      expect(getResp.statusCode).toBe(200);
      expect(getResp.body).toMatchSnapshot();
    });

    it("delete", async () => {
      const patchResp = await request(getServer()).delete("/org/org1/repos/1");
      expect(patchResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org1/repos/1");
      expect(getResp.statusCode).toBe(404);
    });
  });

  describe("Issue", () => {
    beforeAll(async () => {
      await setup();
    });
    afterAll(async () => {
      await destroy();
    });

    it("create", async () => {
      const data = {
        issue: {
          title: "Issue 1",
        },
        c: {
          body: "Comment body",
        },
      };
      const postResp = await request(getServer()).post("/org/org1/repos/1/issues").send(data);
      console.dir(postResp.body);
      expect(postResp.statusCode).toBe(200);

      const getResp = await request(getServer()).get("/org/org1/repos/1/issues/1");
      expect(getResp.statusCode).toBe(200);
      expect(getResp.body).toMatchSnapshot();
    });
  });
});

/** Load definition file and return it's content */
function loadBlueprint(filePath: string): string {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Blueprint file not found: "${filePath}"`);
  }
  return fs.readFileSync(filePath).toString("utf-8");
}

/** Load populator data rom JSON and parse it to object */
function loadPopulatorData(filePath: string): PopulatorData[] {
  try {
    if (!fs.existsSync(filePath)) {
      throw new Error(`Populator data file not found: "${filePath}"`);
    }

    const fileContent = fs.readFileSync(filePath).toString("utf-8");
    return JSON.parse(fileContent);
  } catch (err) {
    if (err instanceof SyntaxError) {
      throw new Error(`Populator data is not valid: ${err.message}`);
    } else {
      throw err;
    }
  }
}
