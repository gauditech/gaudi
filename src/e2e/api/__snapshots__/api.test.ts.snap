// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API endpoints Issue create 1`] = `
{
  "comments": [
    {
      "body": "Comment body",
      "id": 1,
      "issue_id": 1,
    },
  ],
  "id": 1,
  "number": 1,
  "repo": [
    {
      "description": "Repo 1 description",
      "id": 1,
      "is_public": true,
      "latest_num": 1,
      "name": "Repo 1",
      "org_id": 1,
      "slug": "repo1",
    },
  ],
  "title": "Issue 1",
}
`;

exports[`API endpoints Org create 1`] = `
{
  "description": "Org NEW description",
  "id": 5,
  "name": "Org NEW",
  "nameAndDesc": "Org NEW: Org NEW description",
  "slug": "orgNEW",
  "summary": "Org NEWOrg NEW description",
}
`;

exports[`API endpoints Org custom create 1`] = `
{
  "description": "Org custom NEW description",
  "id": 6,
  "name": "Org Custom NEW",
  "nameAndDesc": "Org Custom NEW: Org custom NEW description",
  "slug": "orgCustomNEW",
  "summary": "Org Custom NEWOrg custom NEW description",
}
`;

exports[`API endpoints Org custom endpoint - fetch action 1`] = `
{
  "name": "Fetch me org 1",
  "repoSlug": "repo1",
}
`;

exports[`API endpoints Org custom one endpoint - action with query 1`] = `
{
  "inputName": "Org 1",
  "inputOrgId": 1,
  "orgById": [
    {
      "id": 1,
      "name": "Org 1",
    },
  ],
  "orgByName": [
    {
      "id": 1,
      "name": "Org 1",
    },
  ],
}
`;

exports[`API endpoints Org custom update 1`] = `
{
  "description": "Org custom 2A description",
  "id": 2,
  "name": "Org custom 2A",
  "nameAndDesc": "Org custom 2A: Org custom 2A description",
  "slug": "org2",
  "summary": "Org custom 2AOrg custom 2A description",
}
`;

exports[`API endpoints Org get 1`] = `
{
  "description": "Org 1 description",
  "id": 1,
  "name": "Org 1",
  "nameAndDesc": "Org 1: Org 1 description",
  "slug": "org1",
  "summary": "Org 1Org 1 description",
}
`;

exports[`API endpoints Org list 1`] = `
{
  "data": [
    {
      "description": "Org 1 description",
      "id": 1,
      "name": "Org 1",
      "nameAndDesc": "Org 1: Org 1 description",
      "slug": "org1",
      "summary": "Org 1Org 1 description",
    },
    {
      "description": "Org 2 description",
      "id": 2,
      "name": "Org 2",
      "nameAndDesc": "Org 2: Org 2 description",
      "slug": "org2",
      "summary": "Org 2Org 2 description",
    },
    {
      "description": "Org 3 description",
      "id": 3,
      "name": "Org 3",
      "nameAndDesc": "Org 3: Org 3 description",
      "slug": "org3",
      "summary": "Org 3Org 3 description",
    },
    {
      "description": "Org 4 description",
      "id": 4,
      "name": "Org 4",
      "nameAndDesc": "Org 4: Org 4 description",
      "slug": "org4",
      "summary": "Org 4Org 4 description",
    },
  ],
  "page": 1,
  "pageSize": 20,
  "totalCount": 4,
  "totalPages": 1,
}
`;

exports[`API endpoints Org list with paging 1`] = `
{
  "data": [
    {
      "description": "Org 3 description",
      "id": 3,
      "name": "Org 3",
      "nameAndDesc": "Org 3: Org 3 description",
      "slug": "org3",
      "summary": "Org 3Org 3 description",
    },
    {
      "description": "Org 4 description",
      "id": 4,
      "name": "Org 4",
      "nameAndDesc": "Org 4: Org 4 description",
      "slug": "org4",
      "summary": "Org 4Org 4 description",
    },
  ],
  "page": 2,
  "pageSize": 2,
  "totalCount": 4,
  "totalPages": 2,
}
`;

exports[`API endpoints Org update 1`] = `
{
  "description": "Org 2A description",
  "id": 2,
  "name": "Org 2A",
  "nameAndDesc": "Org 2A: Org 2A description",
  "slug": "org2",
  "summary": "Org 2AOrg 2A description",
}
`;

exports[`API endpoints Repo create 1`] = `
{
  "description": "Description: Repo 6 description",
  "id": 6,
  "org_id": 1,
  "slug": "org_1-W5JU1e0Kj1Kv",
}
`;

exports[`API endpoints Repo get 1`] = `
{
  "description": "Repo 1 description",
  "id": 1,
  "org_id": 1,
  "slug": "repo1",
}
`;

exports[`API endpoints Repo list 1`] = `
{
  "data": [
    {
      "description": "Repo 1 description",
      "id": 1,
      "org_id": 1,
      "slug": "repo1",
    },
    {
      "description": "Repo 2 description",
      "id": 2,
      "org_id": 1,
      "slug": "repo2",
    },
    {
      "description": "Repo 3 description",
      "id": 3,
      "org_id": 1,
      "slug": "repo3",
    },
  ],
  "page": 1,
  "pageSize": 3,
  "totalCount": 3,
  "totalPages": 1,
}
`;

exports[`API endpoints Repo update 1`] = `
{
  "description": "Repo 2A description",
  "id": 2,
  "org_id": 1,
  "slug": "repo2",
}
`;
